lib/membrane_webm/parser/ebml.ex

```diff
@@ -0,0 +1,158 @@
+  EBML Element Data Size VINTs with VINT_DATA consisting of only 1's are reserver to mean `unknown` e.g.:
+    1 1111111
+    0 1 11111111111111
+  determining where an unknonw-sized elements end is tricky
+  https://www.rfc-editor.org/rfc/rfc8794.html#section-6.2
+  """
+  use Bitwise
+
+  alias Membrane.WebM.Schema
+
+  @doc "left for reference but shouldn't be used"
+  def parse_vint(<<first_byte::unsigned-size(8), _rest::binary>> = bytes) do
+    vint_width = get_vint_width(first_byte)
+    <<vint_bytes::binary-size(vint_width), rest::binary>> = bytes
+    <<vint::integer-size(vint_width)-unit(8)>> = vint_bytes
+    vint_data = get_vint_data(vint, vint_width)
+    element_id = Integer.to_string(vint, 16)
+
+    %{
+      vint: %{
+        vint: vint,
+        vint_width: vint_width,
+        vint_data: vint_data,
+        element_id: element_id
+      },
+      rest: rest
+    }
+  end
+
+  @doc """
+  Returns an EBML element's name, type and data
+  """
+  def decode_element(bytes) do
+    {id, bytes} = decode_element_id(bytes)
+    {data_size, bytes} = decode_vint(bytes)
+    name = Schema.element_id_to_name(id)
+    type = Schema.element_type(name)
+
+    # `Segment` is a special element requiring different behaviour of the parser
+    # the parser should only wait for more bytes in case of top-level children elements of `Segment`, not the `Segment` itself
+    if name == :Segment do
+      {:ignore_element_header, bytes}
+    else
+      case split_bytes(bytes, data_size) do
+        {data, bytes} ->
+          # TODO: remove; only for debugging
+          if name == :Unknown do
+            IO.warn("unknown element ID: #{id}")
+          end
+
+          {name, type, data, bytes}
+
+        :need_more_bytes ->
+          :need_more_bytes
+      end
+    end
+  end
+
+  defp split_bytes(bytes, how_many) do
+    if how_many > byte_size(bytes) do
+      :need_more_bytes
+    else
+      <<bytes::binary-size(how_many), rest::binary>> = bytes
+      {bytes, rest}
+    end
+  end
+
+  @doc """
+  Returns the `EBML Element ID` of the given VINT.
+
+  EMBL elements are identified by the hex representation of the entire leading VINT including WIDTH, MARKER and DATA
+  """
+  def decode_element_id(<<first_byte::unsigned-size(8), _rest::binary>> = bytes) do
+    vint_width = get_vint_width(first_byte)
+    <<vint_bytes::binary-size(vint_width), rest::binary>> = bytes
+    <<vint::integer-size(vint_width)-unit(8)>> = vint_bytes
+
+    {Integer.to_string(vint, 16), rest}
+  end
+
+  @doc "Returns the number encoded in the VINT_DATA field of the VINT"
+  def decode_vint(<<first_byte::unsigned-size(8), _rest::binary>> = bytes) do
+    vint_width = get_vint_width(first_byte)
+    <<vint_bytes::binary-size(vint_width), rest::binary>> = bytes
+    <<vint::integer-size(vint_width)-unit(8)>> = vint_bytes
+
+    {get_vint_data(vint, vint_width), rest}
+  end
+
+  # the numbers are bit masks for extracting the data part of a VINT
```

Comment:
can we use hex values for that? It will probably will look more friendly than those huge numbers.